src/pymorize/units.py

```diff
@@ -0,0 +1,108 @@
+
+In case of missing units in either model files or CMIP Tables, this module can not convert from a dimentionless base to something with dimension. Dealing with such thing have to done with `action` section in the Rules module on a per variable basis.
+
+"""
+
+
+# TODO: decide a place to hold the missing unit definitions that pint is not aware of
+# from pathlib import Path
+# _unitsfile = Path(__file__).parent / "units_en.txt"
+# u = pint.UnitRegistry(_unitsfile)
+
+
+ureg = pint.UnitRegistry()
+# ureg.define('degC = degree_Celsius')
+# https://ncics.org/portfolio/other-resources/udunits2/
+ureg.define("degrees_east = deg")
+ureg.define("degree_east = deg")
+ureg.define("degrees_north = deg")
+ureg.define("degree_north = deg")
+ureg.define("degrees_west = -1 * deg")
+ureg.define("degrees_south = -1 * deg")
+# https://planetcalc.com/6777/
+ureg.define("molN = 14.007 * g")
+ureg.define("molC = 12.0107 * g")
+ureg.define("molFe = 55.874 * g")
+
+
+def format_exponent_notation(s, pattern=re.compile(r"(?P<name>\w+)-(?P<exp>\d+)")):
+    "m-2 -> m^-2"
+
+    def correction(match):
+        try:
+            float(match.group())
+        except ValueError:
+            d = match.groupdict()
+            s = f"{d['name']}^-{d['exp']}"
+            return s
+        return match.group()
+
+    return re.sub(pattern, correction, s)
+
+
+def format_power_notation(s, pattern=re.compile(r"(?P<name>\w+)(?P<exp>\d+)")):
+    "m2 -> m^2"
+
+    def correction(match):
+        try:
+            float(match.group())
+        except ValueError:
+            d = match.groupdict()
+            s = "{0[name]}^{0[exp]}".format(d)
+            if d["exp"] == 1:
+                s = "{0[name]}".format(d)
+            return s
+        return match.group()
+
+    return re.sub(pattern, correction, s)
+
+
+def to_caret_notation(unit):
+    "Formats the unit so Pint can understand them"
+    return format_power_notation(format_exponent_notation(unit))
+
+
+def calculate_unit_conversion_factor(a: str, b: str) -> float:
+    """
+    Returns the factor required to convert from unit "a" to unit "b"
+    """
+    # print(a, b)
+    try:
+        A = ureg(a)
+    except (pint.errors.DimensionalityError, pint.errors.UndefinedUnitError):
+        A = to_caret_notation(a)
+        A = ureg(A)
+    try:
+        B = ureg(b)
+    except (pint.errors.DimensionalityError, pint.errors.UndefinedUnitError):
+        B = to_caret_notation(b)
+        B = ureg(B)
+    print(A, B)
+    return A.to(B).magnitude
+
+
+def is_equal(a: str, b: str):
+    "check if both 'a' and 'b' are equal"
+    return ureg(to_caret_notation(a)) == ureg(to_caret_notation(b))
+
+
+def _quicktest():
+    a = 1 * ureg.mmolC
+    b = 1 * ureg.kg
+
+    aa = 1 * ureg.mmolC / (ureg.m * ureg.m) / ureg.d
+    bb = 1 * ureg.kg / (ureg.m * ureg.m) / ureg.s
+
+    print(a.to(b))
+
+    print(aa.to(bb))
+
+    r = calculate_unit_conversion_factor("mmolC/m2/d", "kg m-2 s-1")
+    print(r)
```

Comment:
What is the purpose here? If this is for checking, make it a unit test. For interactive learning, remove the code afterwards for maintainability.

Reply:
This is a temporary thing. I will remove it.